crates/sui-graphql-rpc/src/types/transaction_block_effects.rs

```diff
@@ -105,19 +115,87 @@ impl TransactionBlockEffects {
+    /// possible, otherwise it will fall back to simply displaying the Move abort code and location.
+    async fn errors(&self, ctx: &Context<'_>) -> Result<Option<String>> {
+        let mut status = self.native().status().clone();
+        let resolver: &PackageResolver = ctx.data_unchecked();
+
+        if let NativeExecutionStatus::Failure {
+            error:
+                ExecutionFailureStatus::MoveAbort(MoveLocation { module, .. }, _)
+                | ExecutionFailureStatus::MovePrimitiveRuntimeError(MoveLocationOpt(Some(MoveLocation {
+                    module,
+                    ..
+                }))),
+            command: Some(command_idx),
+        } = &mut status
+        {
+            // Get the Move call that this error is associated with.
+            if let Some(Command::MoveCall(ptb_call)) = self
+                .programmable_transaction()?
+                .and_then(|ptb| ptb.commands.into_iter().nth(*command_idx))
+            {
+                let module_new = module.clone();
+                // Resolve the runtime module ID in the Move abort to the storage ID of the package
+                // that the abort occured in. This is important to make sure that we look at the
+                // correct version of the module when resolving the error.
+                *module = resolver
+                    .resolve_module_id(module_new, ptb_call.package.into())
+                    .await
+                    .map_err(|e| Error::Internal(format!("Error resolving Move location: {e}")))?;
+            }
+        }
+
+        match status {
+            NativeExecutionStatus::Success => Ok(None),
 
             NativeExecutionStatus::Failure {
                 error,
                 command: None,
-            } => Some(error.to_string()),
+            } => Ok(Some(error.to_string())),
 
             NativeExecutionStatus::Failure {
                 error,
                 command: Some(command),
             } => {
+                let error = 'error: {
+                    let ExecutionFailureStatus::MoveAbort(loc, code) = &error else {
+                        break 'error error.to_string();
+                    };
+                    let fname_string = if let Some(fname) = &loc.function_name {
+                        format!(" in function '{}' ", fname)
+                    } else {
+                        " ".to_string()
+                    };
+                    match resolver
+                        .resolve_clever_error(loc.module.clone(), *code)
+                        .await
+                    {
+                        Some(CleverError::CompleteError {
+                            module_id,
+                            error_constant,
+                            source_line_number,
+                            error_identifier,
+                        }) => {
+                            let const_str = error_constant.unwrap_or_else(FBase64::encode);
```

Comment:
This trickery with `Result` probably warrants a comment.

Reply:
Trickery removed in favor of the enum approach you mention below :)